design/deploy-steps.md

```diff
@@ -0,0 +1,134 @@
+<!--
+ This work is licensed under a Creative Commons Attribution 3.0
+ Unported License.
+
+ http://creativecommons.org/licenses/by/3.0/legalcode
+-->
+
+# Customizable deployment procedure
+
+## Status
+
+One of: implementable
+
+## Summary
+
+This design proposes customizing the deployment procedure by providing a high
+level abstraction around Ironic [deploy steps][deploy-steps], as well as
+allowing per-host deployment (IPA) images.
+
+[deploy-steps]: https://docs.openstack.org/ironic/latest/admin/node-deployment.html#deploy-steps
+
+## Motivation
+
+By default the Ironic agent downloads a provided image via HTTP(s), converts it
+(if needed) and writes directly to the disk. This procedure works for any
+distributions, but may not suit some cases well. A prominent example is CoreOS,
+which has its own installer (`coreos-installer`) that allows more
+customization.
+
+### Goals
+
+- A way to execute custom actions during deployment.
+- A way to replace the default deploy procedure with the custom actions.
+- A way to provide a link to the Ironic agent image per host.
+
+### Non-Goals
+
+- Expose the full functionality of Ironic deploy steps.
+- Building custom agent images.
+
+## Proposal
+
+### User Stories
+
+#### Story 1
+
+As a CoreOS user I want to use `coreos-installer` to write the CoreOS image
+without losing any functionality that Ironic offers (unlike the live ISO
+workflow).
+
+#### Story 2
+
+As an operator I want to execute site-specific actions during deployment. I'm
+ready to provide them as plugins to the Ironic agent inside my custom agent
+image.
+
+## Design Details
+
+Additions to `BareMetalHost`:
+
+- New field `AgentImageURL`. It will accept an optional link to a custom agent
+  ISO image.
```

Comment:
We have an `Image` type [1]. We try to be consistent when we can elsewhere in the API. Would it make sense to reuse that here? So, an `agentImage` field in the API would have all of the sub-fields that `image` has today? Or maybe we would define a new type with a subset of those fields, if we don't need something like `diskFormat` for now? Even if that just gives us `image` with `url` as a sub-field, that would open the way to more flexibility later.

[1] https://github.com/metal3-io/baremetal-operator/blob/master/apis/metal3.io/v1alpha1/baremetalhost_types.go#L396

Reply:
Makes sense, especially since we may add kernel/initramfs later.